pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TopicsBase.java

```diff
@@ -0,0 +1,599 @@
+        // Only need to add to first partition the broker owns since the schema id in schema registry are
+        // same for all partitions which is the partitionedTopicName
+        List<Integer> partitions = owningTopics.get(topicName.getPartitionedTopicName()).values();
+        CompletableFuture<SchemaVersion> result = new CompletableFuture<>();
+        for (int index = 0; index < partitions.size(); index++) {
+            CompletableFuture<SchemaVersion> future = new CompletableFuture<>();
+            String topicPartitionName = topicName.getPartition(partitions.get(index)).toString();
+            pulsar().getBrokerService().getTopic(topicPartitionName, false)
+            .thenAccept(topic -> {
+                if (!topic.isPresent()) {
+                    future.completeExceptionally(new BrokerServiceException.TopicNotFoundException(
+                            "Topic " + topicPartitionName + " not found"));
+                } else {
+                    topic.get().addSchema(schemaData).thenAccept(schemaVersion -> future.complete(schemaVersion))
+                    .exceptionally(exception -> {
+                        exception.printStackTrace();
+                        future.completeExceptionally(exception);
+                        return null;
+                    });
+                }
+            });
+            try {
+                result.complete(future.get());
+                break;
+            } catch (Exception e) {
+                result.completeExceptionally(new SchemaException("Unable to add schema " + schemaData
+                        + " to topic " + topicName.getPartitionedTopicName()));
+            }
+        }
+        return result;
+    }
+
+    private SchemaData getSchemaData(String keySchema, String valueSchema) {
+        try {
+            if ((keySchema == null || keySchema.isEmpty()) && (valueSchema == null || valueSchema.isEmpty())) {
+                return null;
+            } else {
+                SchemaInfo keySchemaInfo = (keySchema == null || keySchema.isEmpty())
+                        ? StringSchema.utf8().getSchemaInfo() : ObjectMapperFactory.getThreadLocal()
+                        .readValue(Base64.getDecoder().decode(keySchema), SchemaInfo.class);
+                SchemaInfo valueSchemaInfo = (valueSchema == null || valueSchema.isEmpty())
+                        ? StringSchema.utf8().getSchemaInfo() : ObjectMapperFactory.getThreadLocal()
+                        .readValue(Base64.getDecoder().decode(valueSchema), SchemaInfo.class);
+                if (null == keySchemaInfo.getName()) {
+                    keySchemaInfo.setName(keySchemaInfo.getType().toString());
+                }
+                if (null == valueSchemaInfo.getName()) {
+                    valueSchemaInfo.setName(valueSchemaInfo.getType().toString());
+                }
+                SchemaInfo schemaInfo = KeyValueSchemaInfo.encodeKeyValueSchemaInfo("KVSchema-"
```

Comment:
If there is no key schema, we should just use the value schema. We don't need to use KeyValueSchema.

Reply:
updated.